PowerEditor/src/Notepad_plus.cpp

```diff
@@ -8102,3 +8104,243 @@ void Notepad_plus::updateCommandShortcuts()
+				}
+			}
+
+			HBITMAP hOldBitmap = static_cast<HBITMAP>(::SelectObject(hdcMem, hMarkBitmap));
+
+			SIZE sizeToDraw;
+			if (::GetTextExtentPoint32(hdcMem, pMenuCommandMark->_szMarkStr, nMarkStrLen, &sizeToDraw))
+			{
+				// draw the mark bitmap
+				::FillMemory(pvBits, bmi.bmiHeader.biSizeImage, 0); // black is transparent
+				::SetBkColor(hdcMem, RGB(0, 0, 0)); // black is transparent
+				if (RGB(0, 0, 0) == clrfText)
+					clrfText = RGB(1, 1, 1); // we cannot draw here with the complete black color! (transparency...)
+				::SetTextColor(hdcMem, clrfText);
+				RECT rectMark = { 0, 0, pMenuCommandMark->_sizeMarkArea.cx, pMenuCommandMark->_sizeMarkArea.cy };
+				if (::ExtTextOut(hdcMem, pMenuCommandMark->_sizeMarkArea.cx / 2 - (sizeToDraw.cx + 1) / 2, 0, // try to draw in the middle
+					ETO_OPAQUE | ETO_CLIPPED, &rectMark, pMenuCommandMark->_szMarkStr, nMarkStrLen, NULL))
+				{
+					bMark = true;
+
+					// make all the non-zero pixels opaque
+					LPBYTE pPixBytes = reinterpret_cast<LPBYTE>(pvBits);
+					for (UINT i = 0; i < bmi.bmiHeader.biSizeImage; i += sizeof(DWORD))
+					{
+						DWORD dwPixel = *(reinterpret_cast<LPDWORD>(pPixBytes + i));
+						if (dwPixel != 0)
+							pPixBytes[i + 3] = pMenuCommandMark->_alphaOpacity; // set opacity by the alpha channel of the BGRA pixel
+					}
+				}
+			}
+
+			// restore the old bitmap, so we can delete it later to avoid leak
+			::SelectObject(hdcMem, hOldBitmap);
+		}
+
+		::DeleteDC(hdcMem);
+		hdcMem = NULL;
+	}
+
+	::ReleaseDC(HWND_DESKTOP, hdcDesktop);
+	hdcDesktop = NULL;
+
+	if (!bMark)
+	{
+		if (hMarkBitmap)
+			::DeleteObject(hMarkBitmap);
+		return false;
+	}
+
+	// finally set the mark bitmap for the relevant menu command
+	if (!::SetMenuItemBitmaps(_mainMenuHandle, pMenuCommandMark->_uMenuCommandId, MF_BYCOMMAND, hMarkBitmap, hMarkBitmap))
+	{
+		::DeleteObject(hMarkBitmap);
+		return false;
+	}
+	
+	// optional tooltip part follows
+
+	if (pMenuCommandMark->_szTooltipStrId && pMenuCommandMark->_szTooltipDefaultStr)
+	{
+		if (::IsWindow(pMenuCommandMark->_hTooltip))
+		{
+			::SendMessage(pMenuCommandMark->_hTooltip, TTM_TRACKACTIVATE, FALSE, reinterpret_cast<LPARAM>(&(pMenuCommandMark->_tiTooltip)));
+			::SendMessage(pMenuCommandMark->_hTooltip, TTM_DELTOOL, 0, reinterpret_cast<LPARAM>(&(pMenuCommandMark->_tiTooltip)));
+			::DestroyWindow(pMenuCommandMark->_hTooltip);
+			pMenuCommandMark->_hTooltip = NULL;
+		}
+
+		NativeLangSpeaker* pNativeSpeaker = (NppParameters::getInstance()).getNativeLangSpeaker();
+		if (!pNativeSpeaker)
+			return false;
+		generic_string tip2show = pNativeSpeaker->getLocalizedStrFromID(pMenuCommandMark->_szTooltipStrId, pMenuCommandMark->_szTooltipDefaultStr);
+		TCHAR szBuf[2048]; // we cannot use the tip2show.c_str() directly later
+		_sntprintf_s(szBuf, _countof(szBuf), _TRUNCATE, TEXT("%s"), tip2show.c_str());
+
+		pMenuCommandMark->_hTooltip = CreateWindowEx(pNativeSpeaker->isRTL() ? WS_EX_LAYOUTRTL : 0, TOOLTIPS_CLASS, NULL,

```

Comment:
Since the intent of this PR is the code be part of Notepad++, the reuse of Notepad++ is necessary.
Using the function `CreateToolTip` of Notepad++ could make this PR shorter and more readable: 
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/MISC/Common/Common.cpp#L1015